src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -162,77 +157,124 @@ contract TokenStakingNode is
+            uint256 _shares = withdrawals[i].shares[0];
+            queuedShares[_strategies[i]] -= _shares;
+            withdrawn[_token] += _actualAmountOut;
+
+            if (updateTokenStakingNodesBalances) {
+                IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager())
+                    .updateTokenStakingNodesBalances(_token);
+            }
         }
 
-        emit CompletedQueuedWithdrawals(_shares, _actualAmountOut, address(_strategy));
+        emit CompletedManyQueuedWithdrawals(withdrawals);
+    }
+
+    /**
+     * @notice Completes queued withdrawals with receiveAsTokens set to true
+     * @param withdrawal The withdrawal to complete
+     * @param middlewareTimesIndex The middleware times index
+     * @param updateTokenStakingNodesBalances If true calls updateTokenStakingNodesBalances for yieldNestStrategyManager
+     */
+    function completeQueuedWithdrawals(
+        IDelegationManager.Withdrawal calldata withdrawal,
+        uint256 middlewareTimesIndex,
+        bool updateTokenStakingNodesBalances
+    ) public onlyTokenStakingNodesWithdrawer onlyWhenSynchronized {
+        IDelegationManager.Withdrawal[] memory _withdrawals = new IDelegationManager.Withdrawal[](1);
+        _withdrawals[0] = withdrawal;
+        uint256[] memory _middlewareTimesIndexes = new uint256[](1);
+        _middlewareTimesIndexes[0] = middlewareTimesIndex;
+        completeQueuedWithdrawals(_withdrawals, _middlewareTimesIndexes, updateTokenStakingNodesBalances);
+    }
+
+    /**
+     * @notice Struct containing strategy and shares information
+     * @param strategy The strategy contract address
+     * @param shares The number of shares
+     */
+    struct StrategyShares {
+        IStrategy strategy;
+        uint256 shares;
+    }
+
+    /**
+     * @notice Completes queued withdrawals with receiveAsTokens set to false
+     * @param withdrawals Array of withdrawals to complete
+     * @param middlewareTimesIndexes Array of middleware times indexes
+     * @param strategies Array of strategies to withdraw from
+     */
+    function completeQueuedWithdrawalsAsShares(
+        IDelegationManager.Withdrawal[] calldata withdrawals,
+        uint256[] calldata middlewareTimesIndexes,
+        IStrategy[] calldata strategies
+    ) external onlyDelegator onlyWhenSynchronized {
+        if (withdrawals.length != middlewareTimesIndexes.length || withdrawals.length != strategies.length) {
+            revert ArrayLengthMismatch();
+        }
+
+        IDelegationManager delegationManager = IDelegationManager(address(tokenStakingNodesManager.delegationManager()));
+        IERC20[][] memory _tokens = new IERC20[][](withdrawals.length);
+        bool[] memory _receiveAsTokens = new bool[](withdrawals.length);
+        // Decrease queued shares for each strategy
+        for (uint256 i = 0; i < withdrawals.length; i++) {
+            queuedShares[strategies[i]] -= withdrawals[i].shares[0];
```

Comment:
there's also a problem here with the assumption that there's exactly one strategy queued up

it's not actually checked it's just assumed

if a queued withdrawal is created with more than 1 strategy this  breaks accounting

Let's assert withdrawals[i].shares.length is 1 and so is withdrawals.[i].strategies.length is 1